File: applications/frontend/shared/src/domains/articles/event.ts
import { ulid } from "ulid";
import { Event } from "../common";
import { ArticleIdentifier, ArticleSnapshot } from "./common";

export type ArticleCreatedEvent = Event<
  "article.created",
  {
    snapshot: ArticleSnapshot;
  }
>;

export const createArticleCreatedEvent = (
  article: ArticleSnapshot,
  occurredAt: Date,
): ArticleCreatedEvent => ({
  identifier: ulid(),
  type: "article.created",
  occurredAt,
  payload: {
    snapshot: article,
  },
});

export type ArticleEditedEvent = Event<
  "article.edited",
  {
    next: ArticleSnapshot;
    before: ArticleSnapshot;
  }
>;

export const createArticleEditedEvent = (
  next: ArticleSnapshot,
  before: ArticleSnapshot,
  occurredAt: Date,
): ArticleEditedEvent => ({
  identifier: ulid(),
  type: "article.edited",
  occurredAt,
  payload: {
    next,
    before,
  },
});

export type ArticleTerminatedEvent = Event<
  "article.terminated",
  {
    article: ArticleIdentifier;
  }
>;

export const createArticleTerminatedEvent = (
  article: ArticleIdentifier,
): ArticleTerminatedEvent => ({
  identifier: ulid(),
  type: "article.terminated",
  occurredAt: new Date(),
  payload: {
    article,
  },
});

export type ArticleEvent =
  | ArticleCreatedEvent
  | ArticleEditedEvent
  | ArticleTerminatedEvent;

File: applications/search-token-worker/test/Support/Helper/Domain/Event.hs
module Support.Helper.Domain.Event
  ( createArticleCreatedEvent,
    createArticleEditedEvent,
    createArticleTerminateEvent,
    createMemoCreatedEvent,
    createMemoEditedEvent,
    createMemoTerminateEvent,
  )
where

import Data.Time (UTCTime, addUTCTime, secondsToNominalDiffTime)
import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
import Domain.Common (Timeline (..))
import Domain.Event
  ( ArticleCreatedPayload (..),
    ArticleEditedPayload (..),
    Event (..),
    EventPayload (..),
    EventType (..),
    MemoCreatedPayload (..),
    MemoEditedPayload (..),
    MemoEntry (..),
  )

seedToTime :: Int -> UTCTime
seedToTime seed =
  addUTCTime (secondsToNominalDiffTime (fromIntegral seed)) (posixSecondsToUTCTime 0)

seedToTimeline :: Int -> Timeline
seedToTimeline seed = Timeline (seedToTime seed) (seedToTime (seed + 1))

createEvent :: Int -> EventType -> EventPayload -> Event
createEvent seed eventType' payload' =
  Event
    { identifier = "evt-" <> show seed,
      occurredAt = seedToTime seed,
      eventType = eventType',
      payload = payload'
    }

createArticleCreatedPayload :: Int -> ArticleCreatedPayload
createArticleCreatedPayload seed =
  ArticleCreatedPayload
    { identifier = "article-" <> show seed,
      title = "title-" <> show seed,
      content = "content-" <> show seed,
      excerpt = "excerpt-" <> show seed,
      slug = "slug-" <> show seed,
      status = "published",
      tags = ["tag-" <> show seed],
      timeline = seedToTimeline seed
    }

createMemoCreatedPayload :: Int -> MemoCreatedPayload
createMemoCreatedPayload seed =
  MemoCreatedPayload
    { identifier = "memo-" <> show seed,
      title = "title-" <> show seed,
      slug = "slug-" <> show seed,
      entries = [MemoEntry {text = "entry-" <> show seed, createdAt = seedToTime seed}],
      tags = ["tag-" <> show seed],
      status = "published",
      timeline = seedToTimeline seed
    }

createArticleCreatedEvent :: Int -> Event
createArticleCreatedEvent seed =
  createEvent seed ArticleCreated (ArticleCreatedPayload' (createArticleCreatedPayload seed))

createArticleEditedEvent :: Int -> Event
createArticleEditedEvent seed =
  createEvent
    seed
    ArticleEdited
    ( ArticleEditedPayload'
        ArticleEditedPayload
          { next = createArticleCreatedPayload (seed + 1),
            before = createArticleCreatedPayload seed
          }
    )

createArticleTerminateEvent :: Int -> Event
createArticleTerminateEvent seed =
  createEvent seed ArticleTerminated (ArticleTerminatePayload' ("article-" <> show seed))

createMemoCreatedEvent :: Int -> Event
createMemoCreatedEvent seed =
  createEvent seed MemoCreated (MemoCreatedPayload' (createMemoCreatedPayload seed))

createMemoEditedEvent :: Int -> Event
createMemoEditedEvent seed =
  createEvent
    seed
    MemoEdited
    ( MemoEditedPayload'
        MemoEditedPayload
          { next = createMemoCreatedPayload (seed + 1),
            before = createMemoCreatedPayload seed
          }
    )

createMemoTerminateEvent :: Int -> Event
createMemoTerminateEvent seed =
  createEvent seed MemoTerminated (MemoTerminatePayload' ("memo-" <> show seed))
